iris-mpc-cpu/src/execution/hawk_main.rs

```diff
@@ -362,6 +370,86 @@ impl HawkActor {
         Ok(connect_plan)
     }
 
+    async fn calculate_threshold_a(n_buckets: usize) -> Vec<u32> {
+        (1..=n_buckets)
+            .map(|x: usize| {
+                translate_threshold_a(MATCH_THRESHOLD_RATIO / (n_buckets as f64) * (x as f64))
+            })
+            .collect::<Vec<_>>()
+    }
+
+    pub async fn cache_distances(
+        &mut self,
+        side: usize,
+        search_results: Iter<'_, VecRots<InsertPlan>>,
+    ) -> Result<()> {
+        let mut neighbors: Vec<&SortedNeighborhoodV<Aby3Store>> = vec![];
+        for result in search_results {
+            let n = result
+                .iter()
+                .flat_map(|plan| plan.links.iter())
+                .collect_vec();
+            neighbors.extend(n);
+        }
+
+        for neighbor in neighbors.iter() {
+            let distances = neighbor.distances_cloned();
+            self.distances_cache[side].extend(distances);
+        }
+        Ok(())
+    }
+
+    pub async fn compute_buckets(
+        &self,
+        session: &mut HawkSession,
+        side: usize,
+    ) -> Result<VecBuckets> {
+        let translated_thresholds = Self::calculate_threshold_a(self.args.n_buckets).await;
+        let bucket_result_shares = compare_threshold_buckets(
+            &mut session.aby3_store.session,
+            translated_thresholds.as_slice(),
+            self.distances_cache[side].as_slice(),
+        )
+        .await?;
+
+        let buckets = open_ring(&mut session.aby3_store.session, &bucket_result_shares).await?;
+        Ok(buckets)
+    }
+
+    pub async fn fill_anonymized_statistics_buckets(
+        &mut self,
+        session: &mut HawkSession,
+    ) -> Result<()> {
+        if self.distances_cache[LEFT].len() > self.args.match_distances_buffer_size {
```

Comment:
At some point we should probably rename this config option, because it's a bit confusing with this implementation (since the buffer is not hard capped)